File: Competitive_coding/Competitive_coding_in_PYTHON/first.py
class MinStack:
    def __init__(self, capacity):
        self.capacity = capacity
        self.data = []  # List to hold stack elements
        self.minData = []  # List to hold minimum elements
        self.top = -1  # Index of the top element

    def isFull(self):
        return self.top == self.capacity - 1

    def isEmpty(self):
        return self.top == -1

    def push(self, element):
        if self.isFull():
            print("Overflow")
            return

        self.top += 1
        self.data.append(element)

        # Update the min stack
        if self.top == 0 or element <= self.minData[self.top - 1]:
            self.minData.append(element)
        else:
            self.minData.append(self.minData[self.top - 1])

    def pop(self):
        if self.isEmpty():
            print("Underflow")
            return

        popped = self.data.pop()
        self.top -= 1
        self.minData.pop()

    def top_element(self):
        if self.isEmpty():
            print("Stack is empty")
            return -1

        return self.data[self.top]

    def getMin(self):
        if self.isEmpty():
            print("Stack is empty")
            return -1

        return self.minData[self.top]

    def display(self):
        if self.isEmpty():
            print("Stack is empty")
            return

        print("Elements in the stack:", end=" ")
        for i in range(self.top + 1):
            print(self.data[i], end=" ")
        print()

# Main function to demonstrate the MinStack
if __name__ == "__main__":
    minStack = MinStack(5)

    minStack.push(3)
    minStack.push(5)
    minStack.push(2)
    minStack.push(7)
    minStack.push(1)

    minStack.display()
    print("Min element in stack:", minStack.getMin())

    minStack.pop()
    minStack.pop()

    print("Top element in stack:", minStack.top_element())

File: Competitive_coding/Competitive_coding_by_Faculty/first.c
#include<stdio.h>
#include<stdlib.h>
#include<limits.h>

//Define the maximum element in the stack
#define max 5

typedef struct 
{
    int data[max];
    int min_data[max];
    int top;   
}Min_Stack;

void initilize(Min_Stack *stack){
 stack->top = -1;
}

int isFull(Min_Stack *stack){
    return stack->top == max-1;
}

int isEmpty(Min_Stack *stack){
    return stack->top == -1;
}

void push(Min_Stack *stack , int data){
    if (isFull(stack)){
        printf("Stack is Overflow!");return;
    }
    stack->top++;
    stack->data[stack->top] = data;

    if (stack->top == 0 || data <= stack->min_data[stack->top - 1]) { 
        stack->min_data[stack->top] = data; 
    } else { 
        stack->min_data[stack->top] = stack->min_data[stack->top - 1]; 
    } 
} 

void pop(Min_Stack *stack){
    if (isEmpty(stack)){
        printf("Stack in UnderFlow");
        return;
    }
    int popped_element = stack->data[stack->top];
    stack->top--;
    if(popped_element == stack->min_data[stack->top + 1]){
        stack->min_data[stack->top+1] = INT_MAX;
    }
}

int peek(Min_Stack *stack){
    if (isEmpty(stack)){
        printf("Stack in underflow!");
        return -1;
    }
    return stack->data[stack->top];
}

int getMin(Min_Stack *stack){
    if (isEmpty(stack)){
        printf("Stack is UnderFlow!");
        return -1;
    }
    return stack->min_data[stack->top];
}

void display(Min_Stack *stack){
    if (isEmpty(stack)){
        printf("Stack is UnderFlow!");
        return;
    }
    printf("Elements in Array are: ");
    for (int i = 0; i <= stack->top; i++)
    {
        printf("%d ",stack->data[i]);
    }
    printf("\n");
    
}

int main() { 
    Min_Stack minStack; 
    initilize(&minStack); 
 
    push(&minStack, 3); 
    push(&minStack, 5); 
    push(&minStack, 2); 
    push(&minStack, 7); 
    push(&minStack, 1); 
 
    display(&minStack); 
    printf("Min element in stack: %d\n", getMin(&minStack));  
    
 
    printf("Top element in stack: %d\n", peek(&minStack)); 
     
    return 0; 
} 
